lib/iterator.js

```diff
@@ -0,0 +1,356 @@
+  if (this._map) nodes = nodes.map(this._map)
+  if (this._reduce) return nodes.reduce(this._reduce)
+  return nodes
+}
+
+Iterator.prototype._consume = function (cb) {
+  if (this._updated) {
+    this._updated = false
+    this._next(cb)
+    return
+  }
+
+  if (this._workers.length === 1) {
+    if (this._workers[0].ended) return cb(null, null)
+    var node = consume(this._workers[0])
+    if (node.value === null) return this._next(cb)
+    // TODO: fast case for single node
+    cb(null, this._prereturn([node]))
+    return
+  }
+
+  var min = minKey(this._workers)
+  var nodes = []
+
+  for (var i = 0; i < this._workers.length; i++) {
+    var w = this._workers[i]
+    if (w.value && w.value.key === min.key) nodes.push(consume(w))
+  }
+
+  if (!nodes.length) return cb(null, null)
+  if (allDeletes(nodes)) return this._next(cb)
+  cb(null, this._prereturn(nodes))
+}
+
+Iterator.prototype._fork = function (nodes, from, offset) {
+  if (!nodes.length) return
+
+  if (from) {
+    from.forks = nodes.length
+    var idx = this._workers.indexOf(from)
+    if (idx > -1) this._workers.splice(idx, 1)
+  }
+
+  for (var i = 0; i < nodes.length; i++) {
+    var w = new Worker(nodes[i], offset, from)
+    this._workers.push(w)
+    this._updated = true
+    if (!this._isHead(w.lock, w)) this._endWorker(w, true)
+  }
+}
+
+Iterator.prototype._endWorker = function (worker) {
+  if (worker.from) this._endFork(worker)
+
+  if (worker.notHead) {
+    var i = this._workers.indexOf(worker)
+    this._workers.splice(i, 1)
+    this._updated = true
+  }
+}
+
+Iterator.prototype._endFork = function (worker) {
+  if (--worker.from.forks) return
+  this._workers.push(worker.from)
+  this._updated = true
+
+  var updated = true
+  while (updated) {
+    updated = false
+    for (var i = 0; i < this._workers.length; i++) {
+      if (this._workers[i].from === worker.from) {
+        this._workers.splice(i, 1)
+        updated = true
+        break
+      }
+    }
+  }
+}
+
+Iterator.prototype._isHead = function (head, worker) {
+  var clock = head.seq + 1
+
+  for (var i = 0; i < this._workers.length; i++) {
+    var otherWorker = this._workers[i]
+    if (otherWorker === worker) continue
+
+    var otherClock = otherWorker.lock.clock[head.feed]
+    if (clock <= otherClock) return false
+  }
+
+  return true
+}
+
+function Worker (head, i, from) {
+  this.lock = head
+  this.value = null
+  this.forks = 0
+  this.ended = false
+  this.stack = []
+  this.pending = 0
+  this.error = null
+  this.callback = null
+  this.from = from
+
+  this.stack.push({node: head, i, fork: null})
+}
+
+function copyLock (lock) {
+  return {
+    clock: lock.clock.slice(0)
+  }
+}
+
+function consume (worker) {
+  var val = worker.value
+  worker.value = null
+  return val
+}
+
+function wait (worker, cb) {
+  worker.callback = cb
+}
+
+function fork (iterator, worker, nodes, i, cb) {
+  iterator._fork(nodes, worker, i)
+  cb(null)
```

Comment:
`process.nextTick(cb, null)`?

Reply:
this is guaranteed to run in the nexttick already so no need :)